### PR TITLE
Add amd gpgpu tests (New)

### PR DESCRIPTION
**Pull request by pedro-avalos**
## Description

- Added a wrapper for ROCm Validation suite that checks the result of the tests
- Added AMD GPGPU test jobs using ROCm Validation Suite

## Resolved issues

Addresses [CHECKBOX-969](https://warthogs.atlassian.net/browse/CHECKBOX-969) and [CHECKBOX-1539](https://warthogs.atlassian.net/browse/CHECKBOX-1539).

## Documentation

Coverage tests are included with this PR.

## Tests

I tested this on my desktop, which has an AMD GPU (`AMD Radeon RX 6750 XT`). NOTE: For some AMD GPUs, you may need to change the environment variable `HSA_OVERRIDE_GFX_VERSION` for `rvs` to recognize that your GPU is compatible with some ROCm/HIP functions. For example, for my GPU, I had to edit the value to `10.3.0`.

[CHECKBOX-969]: https://warthogs.atlassian.net/browse/CHECKBOX-969?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CHECKBOX-1539]: https://warthogs.atlassian.net/browse/CHECKBOX-1539?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by pedro-avalos** [comment]
One design question I'd like some feedback on is how `rvs.py` takes the modules as positional arguments. Since we're likely to just use these one at a time in Checkbox jobs, should these be subparsers instead? If they are subparsers, then we could conceivably add the module configuration as options for those subparsers. The more I think about it, they should be subparsers. Let me know what you think!

**Comment by fernando79513** [comment]
> akes the modules as positional arguments. Since we're likely to just use these one at a time in Checkbox jobs, should these be subparsers instead? If they are subparsers, then we could conceivably add the module configuration as options for those subparsers. The more I think about it, they should be subparsers. Let me know what you think!

I think the configs are complex enough to be kept into separated config files. I would restrict the `modules` parameter to run only one module, and I think you could simplify the argparser by setting `choices` for the possible modules:
```
    parser.add_argument(
        "modules",
        metavar="MOD",
        nargs="*",
        choices=RVS_MODULES.keys(),
        type=str,
        help="RVS modules to run [{}]".format(", ".join(RVS_MODULES)),
    )
```

**Review by fernando79513 — CHANGES_REQUESTED** [review]
Some small changes. You don't need to handle the return values if you are not going to use them. Consider just raising a `SystemExit()` with the message you want to simplify the code.

File: providers/gpgpu/bin/rvs.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+    proc = subprocess.run(
+        ["which", "rvs"],
+        check=False,
+        stdout=subprocess.PIPE,
+        universal_newlines=True,
+    )
+    return Path(proc.stdout.strip()) if proc.returncode == 0 else RVS_BIN
+
+
+class ModuleRunner:
```
Comment (fernando79513) [inline providers/gpgpu/bin/rvs.py:50]: I think the whole ModuleRunner class could be simplified and made more explicit, getting rid of all the `subprocess.run` wrappers and the complexity in the validators:
```
class ModuleRunner:
    """This class represents the base module runner."""

    def __init__(self, rvs: Path, config: Path) -> None:
        """Initializes the module runner."""
        self.config = config
        self.rvs = rvs or Path(shutil.which("rvs") or RVS_BIN)

    def run(self, module: str):
        logging.debug("%s: RUNNING", module)

        result = subprocess.run(
            [self.rvs, "-c", self.config],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            universal_newlines=True,
        )

        if result.returncode != 0:
            logging.error(result.stdout)
            logging.error(result.stderr)
            raise SystemExit("%s: FAILURE - failed to execute" % module)

        # Log the error output as debug
        if result.stderr:
            logging.debug(result.stderr)

        if hasattr(self, "_validate_output"):
            if not self._validate_output(result.stdout, module):
                logging.error(result.stdout)
                raise SystemExit("%s: FAILURE - validation failed" % module)
            
        logging.info(result.stdout)
        logging.debug("%s: SUCCESS", module)


class PassFailModuleRunner(ModuleRunner):
    """This class represents a module runner that passes or fails."""

    def _validate_output(self, output: str, module: str):
        # Find any of the common success messages in stdout
        for line in output.splitlines():
            if any(
                re.search(pass_str, line)
                for pass_str in [
                    r"%s true" % re.escape(module),
                    r"pass: TRUE",
                    r"GFLOPS \d+ Target GFLOPS: \d+ met: TRUE",
                ]
            ):
                return False
        return True


class MemModuleRunner(ModuleRunner):
    """This class represents the memory test module runner."""

    def _validate_output(self, output: str, module: str):
        return not all(
            "mem Test %s : PASS" % i in output.splitlines()
            for i in range(1, 12)
        )
```

**Review by fernando79513 — APPROVED** [review]
LGTM +1!

**Pull request merged**